Generic bounds checking fails when bound is not Parcelable but declared type is.

``` java
interface Foo {}
interface Bar<F extends Foo> {
  F foo();
}
@AutoValue abstract class PFoo implements Foo, Parcelable {}
@AutoValue abstract class PBar implements Bar<PFoo>, Parcelable {}
```

```
Bar.java:340: error: AutoValue property foo is not a supported Parcelable type.
    F foo();
      ^
PBar:11: error: @AutoValue processor threw an exception: com.ryanharter.auto.value.parcel.AutoValueParcelException
@AutoValue public abstract class PBar implements Bar<PFoo>,
                           ^
        at com.ryanharter.auto.value.parcel.AutoValueParcelExtension.validateProperties(AutoValueParcelExtension.java:279)
        at com.ryanharter.auto.value.parcel.AutoValueParcelExtension.generateClass(AutoValueParcelExtension.java:158)
        at com.google.auto.value.processor.AutoValueProcessor.writeExtensions(AutoValueProcessor.java:474)
        at com.google.auto.value.processor.AutoValueProcessor.processType(AutoValueProcessor.java:460)
        at com.google.auto.value.processor.AutoValueProcessor.process(AutoValueProcessor.java:150)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:794)
```

Looks like the implementation is resolving the bounds and trying to verify they are parcelable which fails, except that the generic was given a concrete type on the implementation class.

AV 1.3, extension 0.2.5

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -1,0 +1,22 @@
+name: CI
+
+on: [push, pull_request]
+
+jobs:
+  build:
+    name: JDK ${{ matrix.java_version }}
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        java_version: [1.8]
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v1
+      - name: Gradle Wrapper Validation
+        uses: gradle/wrapper-validation-action@v1
+      - name: Install JDK ${{ matrix.java_version }}
+        uses: actions/setup-java@v1
+        with:
+          java-version: ${{ matrix.java_version }}
+      - name: Build with Gradle
+        run: ./gradlew clean check --stacktrace

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,2 +1,0 @@
-language: java
-jdk: openjdk8

--- a/auto-value-parcel/src/main/java/com/ryanharter/auto/value/parcel/AutoValueParcelExtension.java
+++ b/auto-value-parcel/src/main/java/com/ryanharter/auto/value/parcel/AutoValueParcelExtension.java
@@ -63,16 +63,18 @@ public final class AutoValueParcelExtension extends AutoValueExtension {
     final String methodName;
     final String humanName;
     final ExecutableElement element;
+    final TypeMirror typeMirror;
     final TypeName type;
     final ImmutableSet<String> annotations;
     final boolean nullable;
     TypeMirror typeAdapter;
 
-    public Property(String humanName, ExecutableElement element) {
+    public Property(String humanName, ExecutableElement element, TypeMirror actualType) {
       this.methodName = element.getSimpleName().toString();
       this.humanName = humanName;
       this.element = element;
-      type = TypeName.get(element.getReturnType());
+      typeMirror = actualType;
+      type = TypeName.get(actualType);
       annotations = buildAnnotations(element);
       nullable = nullableAnnotation() != null;
 
@@ -175,7 +177,7 @@ public final class AutoValueParcelExtension extends AutoValueExtension {
     TypeName autoValueType =
         TypeName.get(env.getTypeUtils().erasure(context.autoValueClass().asType()));
 
-    ImmutableList<Property> properties = readProperties(context.properties());
+    ImmutableList<Property> properties = readProperties(context);
     if (!validateProperties(env, properties)) {
       return null;
     }
@@ -274,10 +276,11 @@ public final class AutoValueParcelExtension extends AutoValueExtension {
     return null;
   }
 
-  private ImmutableList<Property> readProperties(Map<String, ExecutableElement> properties) {
+  private ImmutableList<Property> readProperties(Context context) {
     ImmutableList.Builder<Property> values = ImmutableList.builder();
-    for (Map.Entry<String, ExecutableElement> entry : properties.entrySet()) {
-      values.add(new Property(entry.getKey(), entry.getValue()));
+    for (Map.Entry<String, ExecutableElement> entry : context.properties().entrySet()) {
+      String name = entry.getKey();
+      values.add(new Property(name, entry.getValue(), context.propertyTypes().get(name)));
     }
     return values.build();
   }
@@ -288,7 +291,7 @@ public final class AutoValueParcelExtension extends AutoValueExtension {
       if (property.typeAdapter != null) {
         continue;
       }
-      TypeMirror type = property.element.getReturnType();
+      TypeMirror type = property.typeMirror;
       if (type.getKind() == TypeKind.ARRAY) {
         ArrayType aType = (ArrayType) type;
         type = aType.getComponentType();

--- a/auto-value-parcel/src/main/java/com/ryanharter/auto/value/parcel/Parcelables.java
+++ b/auto-value-parcel/src/main/java/com/ryanharter/auto/value/parcel/Parcelables.java
@@ -261,7 +261,7 @@ final class Parcelables {
   }
 
   private static boolean isSubclassOf(TypeName type, Types typeUtils, Property property) {
-    TypeMirror clazz = property.element.getReturnType();
+    TypeMirror clazz = property.typeMirror;
 
     while (clazz.getKind() != TypeKind.NONE) {
       TypeName typeName = TypeName.get(clazz);
@@ -285,7 +285,7 @@ final class Parcelables {
 
   private static void readImmutableCollection(CodeBlock.Builder block, Property property,
       boolean isMap) {
-    DeclaredType collectionType = property.element.getReturnType()
+    DeclaredType collectionType = property.typeMirror
         .accept(new SimpleTypeVisitor6<DeclaredType, Void>() {
           @Override
           public DeclaredType visitDeclared(DeclaredType t, Void v) {
@@ -327,8 +327,12 @@ final class Parcelables {
     block.add(expression.toString(), args.toArray());
   }
 
-  public static CodeBlock writeValue(Types types, AutoValueParcelExtension.Property property,
-      ParameterSpec out, ParameterSpec flags) {
+  public static CodeBlock writeValue(
+      Types types,
+      AutoValueParcelExtension.Property property,
+      ParameterSpec out,
+      ParameterSpec flags
+  ) {
     CodeBlock.Builder block = CodeBlock.builder();
 
     TypeName type = getTypeNameFromProperty(property, types);
@@ -445,7 +449,7 @@ final class Parcelables {
   }
 
   static TypeName getTypeNameFromProperty(AutoValueParcelExtension.Property property, Types types) {
-    TypeMirror returnType = property.element.getReturnType();
+    TypeMirror returnType = property.typeMirror;
     if (returnType.getKind() == TypeKind.TYPEVAR) {
       TypeVariable vType = (TypeVariable) returnType;
       returnType = vType.getUpperBound();

--- a/auto-value-parcel/src/test/java/com/ryanharter/auto/value/parcel/AutoValueParcelExtensionTest.java
+++ b/auto-value-parcel/src/test/java/com/ryanharter/auto/value/parcel/AutoValueParcelExtensionTest.java
@@ -6,6 +6,7 @@ import com.google.auto.value.extension.AutoValueExtension;
 import com.google.auto.value.processor.AutoValueProcessor;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Maps;
 import com.google.testing.compile.CompilationRule;
 import com.google.testing.compile.JavaFileObjects;
 import com.ryanharter.auto.value.parcel.util.TestMessager;
@@ -21,6 +22,7 @@ import javax.lang.model.element.ExecutableElement;
 import javax.lang.model.element.Modifier;
 import javax.lang.model.element.TypeElement;
 import javax.lang.model.type.TypeKind;
+import javax.lang.model.type.TypeMirror;
 import javax.lang.model.util.Elements;
 import javax.tools.JavaFileObject;
 import org.junit.Before;
@@ -2169,6 +2171,10 @@ public class AutoValueParcelExtensionTest {
 
     @Override public Set<ExecutableElement> abstractMethods() {
       return Collections.emptySet();
+    }
+
+    @Override public Map<String, TypeMirror> propertyTypes() {
+      return Maps.transformValues(properties, ExecutableElement::getReturnType);
     }
   }
 

--- a/auto-value-parcel/src/test/java/com/ryanharter/auto/value/parcel/AutoValueParcelExtensionTest.java
+++ b/auto-value-parcel/src/test/java/com/ryanharter/auto/value/parcel/AutoValueParcelExtensionTest.java
@@ -215,7 +215,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
         .that(Arrays.asList(parcel, parcelable, nullable, source))
-        .processedWith(new AutoValueProcessor())
+        .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
         .compilesWithoutError()
         .and()
         .generatesSources(expected);
@@ -279,7 +279,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
         .that(Arrays.asList(parcel, parcelable, nullable, source1, source2))
-        .processedWith(new AutoValueProcessor())
+        .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
         .compilesWithoutError()
         .and()
         .generatesSources(expected);
@@ -305,7 +305,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
         .that(Arrays.asList(parcel, parcelable, source))
-        .processedWith(new AutoValueProcessor())
+        .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
         .compilesWithoutError();
   }
 
@@ -408,7 +408,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
         .that(Arrays.asList(parcel, parcelable, notMatching, matching))
-        .processedWith(new AutoValueProcessor())
+        .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
         .compilesWithoutError()
         .and()
         .generatesSources(expectedNotMatching, expectedMatching);
@@ -430,7 +430,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
         .that(Arrays.asList(parcel, parcelable, source))
-        .processedWith(new AutoValueProcessor())
+        .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
         .failsToCompile()
         .withErrorContaining("Manual implementation of Parcelable#writeToParcel(Parcel,int) found when "
                              + "processing test.Test. Remove this so auto-value-parcel can automatically "
@@ -460,7 +460,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
         .that(Arrays.asList(parcel, parcelable, source))
-        .processedWith(new AutoValueProcessor())
+        .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
         .failsToCompile()
         .withErrorContaining("Manual implementation of a static Parcelable.Creator<T> CREATOR field "
                              + "found when processing test.Test. Remove this so auto-value-parcel can "
@@ -939,7 +939,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
         .that(Arrays.asList(nullable, parcel, parcelable, textUtils, parcelable1, foobinder, numbersEnum, parcelableEnum, source))
-        .processedWith(new AutoValueProcessor())
+        .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
         .compilesWithoutError()
         .and()
         .generatesSources(expected);
@@ -1014,7 +1014,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
         .that(Arrays.asList(parcel, parcelable, nullable, source))
-        .processedWith(new AutoValueProcessor())
+        .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
         .compilesWithoutError()
         .and()
         .generatesSources(expected);
@@ -1052,7 +1052,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
         .that(Arrays.asList(parcelable, parcel, source))
-        .processedWith(new AutoValueProcessor())
+        .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
         .failsToCompile()
         .withErrorContaining("Maps can only have String objects for keys and valid Parcelable"
             + " types for values.");
@@ -1143,7 +1143,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
         .that(Arrays.asList(parcel, parcelable, bar, barAdapter, foo))
-        .processedWith(new AutoValueProcessor())
+        .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
         .compilesWithoutError()
         .and()
         .generatesSources(expected);
@@ -1207,7 +1207,7 @@ public class AutoValueParcelExtensionTest {
         + "}");
     assertAbout(javaSources())
         .that(Arrays.asList(parcelable, parcel, source))
-        .processedWith(new AutoValueProcessor())
+        .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
         .compilesWithoutError()
         .and()
         .generatesSources(expected);
@@ -1324,7 +1324,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
         .that(Arrays.asList(parcel, parcelable, bar, baz, foo))
-        .processedWith(new AutoValueProcessor())
+        .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
         .compilesWithoutError()
         .and()
         .generatesSources(expected);
@@ -1420,7 +1420,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
         .that(Arrays.asList(nullable, parcel, parcelable, bar, barAdapter, foo))
-        .processedWith(new AutoValueProcessor())
+        .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
         .compilesWithoutError()
         .and()
         .generatesSources(expected);
@@ -1488,7 +1488,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
         .that(Arrays.asList(parcelable, parcel, source1, source2, source3))
-        .processedWith(new AutoValueProcessor())
+        .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
         .compilesWithoutError()
         .and()
         .generatesSources(expected);
@@ -1549,7 +1549,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
             .that(Arrays.asList(parcel, parcelable, source))
-            .processedWith(new AutoValueProcessor())
+            .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
             .compilesWithoutError()
             .and()
             .generatesSources(expected);
@@ -1610,7 +1610,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
             .that(Arrays.asList(parcel, parcelable, source))
-            .processedWith(new AutoValueProcessor())
+            .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
             .compilesWithoutError()
             .and()
             .generatesSources(expected);
@@ -1670,7 +1670,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
             .that(Arrays.asList(parcel, parcelable, source))
-            .processedWith(new AutoValueProcessor())
+            .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
             .compilesWithoutError()
             .and()
             .generatesSources(expected);
@@ -1794,7 +1794,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
             .that(Arrays.asList(parcel, parcelable, textUtils, source))
-            .processedWith(new AutoValueProcessor())
+            .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
             .compilesWithoutError()
             .and()
             .generatesSources(expected);
@@ -1852,7 +1852,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
             .that(Arrays.asList(parcel, parcelable, source))
-            .processedWith(new AutoValueProcessor())
+            .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
             .compilesWithoutError()
             .and()
             .generatesSources(expected);
@@ -1917,7 +1917,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
             .that(Arrays.asList(parcel, parcelable, source))
-            .processedWith(new AutoValueProcessor())
+            .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
             .compilesWithoutError()
             .and()
             .generatesSources(expected);
@@ -1982,7 +1982,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
         .that(Arrays.asList(parcel, parcelable, source))
-        .processedWith(new AutoValueProcessor())
+        .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
         .compilesWithoutError()
         .and()
         .generatesSources(expected);
@@ -2001,7 +2001,7 @@ public class AutoValueParcelExtensionTest {
 
     assertAbout(javaSources())
             .that(Arrays.asList(parcel, parcelable, source))
-            .processedWith(new AutoValueProcessor())
+            .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
             .failsToCompile()
             .withErrorContaining("AutoValue property tea is not a supported Parcelable type.")
             .in(source).onLine(5);

--- a/auto-value-parcel/src/test/java/com/ryanharter/auto/value/parcel/AutoValueParcelExtensionTest.java
+++ b/auto-value-parcel/src/test/java/com/ryanharter/auto/value/parcel/AutoValueParcelExtensionTest.java
@@ -4,6 +4,7 @@ import android.os.Parcelable;
 import com.google.auto.common.MoreElements;
 import com.google.auto.value.extension.AutoValueExtension;
 import com.google.auto.value.processor.AutoValueProcessor;
+import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 import com.google.testing.compile.CompilationRule;
 import com.google.testing.compile.JavaFileObjects;
@@ -2004,6 +2005,123 @@ public class AutoValueParcelExtensionTest {
             .failsToCompile()
             .withErrorContaining("AutoValue property tea is not a supported Parcelable type.")
             .in(source).onLine(5);
+  }
+
+  @Test
+  public void useResolvedTypesInGenerics() {
+    JavaFileObject foo = JavaFileObjects.forSourceString("test.Foo", ""
+            + "package test;\n"
+            + "interface Foo {}\n"
+    );
+
+    JavaFileObject bar = JavaFileObjects.forSourceString("test.Bar", ""
+            + "package test;\n"
+            + "\n"
+            + "interface Bar<F extends Foo> {\n"
+            + "  F foo();\n"
+            + "}\n"
+    );
+
+    JavaFileObject pfoo = JavaFileObjects.forSourceString("test.PFoo", ""
+            + "package test;\n"
+            + "\n"
+            + "import android.os.Parcelable;"
+            + "import com.google.auto.value.AutoValue;"
+            + "\n"
+            + "@AutoValue abstract class PFoo implements Foo, Parcelable {}\n"
+    );
+
+    JavaFileObject pbar = JavaFileObjects.forSourceString("test.PBar", ""
+            + "package test;\n"
+            + "\n"
+            + "import android.os.Parcelable;"
+            + "import com.google.auto.value.AutoValue;"
+            + "\n"
+            + "@AutoValue abstract class PBar implements Bar<PFoo>, Parcelable {}\n"
+    );
+
+    JavaFileObject expectedPFoo = JavaFileObjects.forSourceString("test.AutoValue_PFoo", "package"
+        + " test;\n"
+        + "\n"
+        + "import android.os.Parcel;\n"
+        + "import android.os.Parcelable;\n"
+        + "import java.lang.Override;\n"
+        + "import javax.annotation.Generated;\n"
+        + "\n"
+        + "@Generated(\"com.ryanharter.auto.value.parcel.AutoValueParcelExtension\")\n"
+        + "final class AutoValue_PFoo extends $AutoValue_PFoo {\n"
+        + "  public static final Parcelable.Creator<AutoValue_PFoo> CREATOR = new Parcelable.Creator<AutoValue_PFoo>() {\n"
+        + "    @Override\n"
+        + "    public AutoValue_PFoo createFromParcel(Parcel in) {\n"
+        + "      return new AutoValue_PFoo(\n"
+        + "      );\n"
+        + "    }\n"
+        + "    @Override\n"
+        + "    public AutoValue_PFoo[] newArray(int size) {\n"
+        + "      return new AutoValue_PFoo[size];\n"
+        + "    }\n"
+        + "  };\n"
+        + "\n"
+        + "  AutoValue_PFoo() {\n"
+        + "    super();\n"
+        + "  }\n"
+        + "\n"
+        + "  @Override\n"
+        + "  public void writeToParcel(Parcel dest, int flags) {\n"
+        + "  }\n"
+        + "\n"
+        + "  @Override\n"
+        + "  public int describeContents() {\n"
+        + "    return 0;\n"
+        + "  }\n"
+        + "}"
+    );
+
+    JavaFileObject expectedPBar = JavaFileObjects.forSourceString("test.AutoValue_PBar", "package"
+        + " test;\n"
+        + "\n"
+        + "import android.os.Parcel;\n"
+        + "import android.os.Parcelable;\n"
+        + "import java.lang.Override;\n"
+        + "import javax.annotation.Generated;\n"
+        + "\n"
+        + "@Generated(\"com.ryanharter.auto.value.parcel.AutoValueParcelExtension\")\n"
+        + "final class AutoValue_PBar extends $AutoValue_PBar {\n"
+        + "  public static final Parcelable.Creator<AutoValue_PBar> CREATOR = new Parcelable.Creator<AutoValue_PBar>() {\n"
+        + "    @Override\n"
+        + "    public AutoValue_PBar createFromParcel(Parcel in) {\n"
+        + "      return new AutoValue_PBar(\n"
+        + "          (PFoo) in.readParcelable(PBar.class.getClassLoader())\n"
+        + "      );\n"
+        + "    }\n"
+        + "    @Override\n"
+        + "    public AutoValue_PBar[] newArray(int size) {\n"
+        + "      return new AutoValue_PBar[size];\n"
+        + "    }\n"
+        + "  };\n"
+        + "\n"
+        + "  AutoValue_PBar(PFoo foo) {\n"
+        + "    super(foo);\n"
+        + "  }\n"
+        + "\n"
+        + "  @Override\n"
+        + "  public void writeToParcel(Parcel dest, int flags) {\n"
+        + "    dest.writeParcelable(foo(), flags);\n"
+        + "  }\n"
+        + "\n"
+        + "  @Override\n"
+        + "  public int describeContents() {\n"
+        + "    return 0;\n"
+        + "  }\n"
+        + "}\n"
+    );
+
+    assertAbout(javaSources())
+        .that(ImmutableList.of(parcel, parcelable, foo, bar, pfoo, pbar))
+        .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
+        .compilesWithoutError()
+        .and()
+        .generatesSources(expectedPFoo, expectedPBar);
   }
 
   private AutoValueExtension.Context createContext(TypeElement type) {

